### PR TITLE
Make truncate and reset safe with concurrent readers

### LINKED ISSUE #4198: SegmentedJournalWriter.truncate() and readers are not thread safe
This is currently not a problem as leaders never truncate. On followers there is no concurrent reader. (We will remove the reader in ZeebeRaftStateMachine) However this can be an issue when we build state on followers #3894

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Use ReadWriteLock for concurrency control between deleteAfter/reset operation by a writer and all operations by readers. It is not necessary for append to acquire lock, as it is does not cause an inconsistent state observed by the reader. As a result, this is read heavy workload because `deleteUntil` and `reset` are only called infrequently.

Blocked by #6814 

## Related issues

closes #4198 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
Just a small note, I think rule of thumb is RWLock have poor performance (unless the read operation is expensive and outweighs the cost of the synchronization overhead), and if you don't need them to be reeantrant, then `StampedLock` over better performance (at the cost of higher complexity). It's unclear to me if it would make a difference to us here, but just a thought (I think StampedLock also offers optimistic locking capabilities, but I guess it's out of the question for our use case)

**Comment by deepthidevaki** [comment]
> Just a small note, I think rule of thumb is RWLock have poor performance (unless the read operation is expensive and outweighs the cost of the synchronization overhead), and if you don't need them to be reeantrant, then `StampedLock` over better performance (at the cost of higher complexity). It's unclear to me if it would make a difference to us here, but just a thought (I think StampedLock also offers optimistic locking capabilities, but I guess it's out of the question for our use case)

I thought StampedLock offers better performance if we can utilize optimistic reading. Does it have better performance even if we are only using normal read and writelock?
I'm currently using the reentrant behavior. However, it is not a requirement. The code can be updated to handle it without reentrancy as it is still internal to journal.

**Comment by npepinpe** [comment]
Honestly I'm no expert, this is knowledge I've mostly gleaned through exposure, not hard testing. As I said, it's been a rule of thumb for a while that RWLock have poor performance and StampedLock behave better (even without using the optimistic variant), but of course I could be wrong in many ways: 1) new JDK could have improved the performance (they've been in since JDK 5 after all), 2) could be in our case RWLock outperform StampedLock as you mention (or because we so seldom truncate/reset), 3) could be StampedLock have similar performance without optimistic (though AFAIK they should outperform even in the pessimistic case, as from what I remember even synchronized outperformed RWLock when read operations were very fast - this may ofc also be outdated knowledge :sweat_smile:)

**Comment by deepthidevaki** [comment]
StampedLock is ~10% faster than ReentrantReadWritelock in a microbenchmark with 4 concurrent journal readers, no writers. `synchronized` is ~400% faster.

```
Benchmark                                      Mode  Cnt        Score        Error  Units
JournalBenchmark.concurrentReadersWithRWLock  thrpt   25  4207523.276 ± 106645.131  ops/s

Benchmark                                           Mode  Cnt        Score       Error  Units
JournalBenchmark.concurrentReadersWithStampedLock  thrpt   25  4724035.260 ± 91578.724  ops/s

Benchmark                                            Mode  Cnt         Score         Error  Units
JournalBenchmark.concurrentReadersWithSynchronized  thrpt   25  20288508.552 ± 2390221.790  ops/s
```
------------------------------------
```
# JMH version: 1.29
# VM version: JDK 11.0.10, OpenJDK 64-Bit Server VM, 11.0.10+9-Ubuntu-0ubuntu1.20.04
# VM invoker: /usr/lib/jvm/java-11-openjdk-amd64/bin/java
# VM options: -Dvisualvm.id=18370673335406 -javaagent:/home/deepthi/.local/share/JetBrains/Toolbox/apps/IDEA-C/ch-0/211.6693.111/lib/idea_rt.jar=35217:/home/deepthi/.local/share/JetBrains/Toolbox/apps/IDEA-C/ch-0/211.6693.111/bin -Dfile.encoding=UTF-8
# Blackhole mode: full + dont-inline hint
# Warmup: 5 iterations, 10 s each
# Measurement: 5 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 4 threads, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.zeebe.journal.file.JournalBenchmark.concurrentReadersWithRWLock

# Run progress: 0.00% complete, ETA 00:08:20
# Fork: 1 of 5
# Warmup Iteration   1: 3901368.520 ops/s
# Warmup Iteration   2: 4079990.392 ops/s
# Warmup Iteration   3: 4021226.240 ops/s
# Warmup Iteration   4: 4095868.259 ops/s
# Warmup Iteration   5: 4131374.195 ops/s
Iteration   1: 4090692.882 ops/s
Iteration   2: 4112894.341 ops/s
Iteration   3: 4196827.879 ops/s
Iteration   4: 4073869.878 ops/s
Iteration   5: 4092571.029 ops/s

# Run progress: 20.00% complete, ETA 00:06:46
# Fork: 2 of 5
# Warmup Iteration   1: 4119722.562 ops/s
# Warmup Iteration   2: 4367213.825 ops/s
# Warmup Iteration   3: 4406530.734 ops/s
# Warmup Iteration   4: 4393575.093 ops/s
# Warmup Iteration   5: 4410656.810 ops/s
Iteration   1: 4411329.720 ops/s
Iteration   2: 4461719.771 ops/s
Iteration   3: 4418016.113 ops/s
Iteration   4: 4421472.350 ops/s
Iteration   5: 4365644.915 ops/s

# Run progress: 40.00% complete, ETA 00:05:03
# Fork: 3 of 5
# Warmup Iteration   1: 3975158.908 ops/s
# Warmup Iteration   2: 4037488.133 ops/s
# Warmup Iteration   3: 4144165.546 ops/s
# Warmup Iteration   4: 4046901.989 ops/s
# Warmup Iteration   5: 4032918.388 ops/s
Iteration   1: 4149272.014 ops/s
Iteration   2: 4056224.601 ops/s
Iteration   3: 4094818.092 ops/s
Iteration   4: 4049753.041 ops/s
Iteration   5: 4042818.270 ops/s

# Run progress: 60.00% complete, ETA 00:03:22
# Fork: 4 of 5
# Warmup Iteration   1: 3904777.796 ops/s
# Warmup Iteration   2: 4076975.066 ops/s
# Warmup Iteration   3: 4078076.382 ops/s
# Warmup Iteration   4: 4059099.932 ops/s
# Warmup Iteration   5: 4169924.588 ops/s
Iteration   1: 4070073.072 ops/s
Iteration   2: 4120999.356 ops/s
Iteration   3: 4092154.898 ops/s
Iteration   4: 4095355.370 ops/s
Iteration   5: 4161045.167 ops/s

# Run progress: 80.00% complete, ETA 00:01:41
# Fork: 5 of 5
# Warmup Iteration   1: 4179024.831 ops/s
# Warmup Iteration   2: 4307303.553 ops/s
# Warmup Iteration   3: 4369555.321 ops/s
# Warmup Iteration   4: 4312779.690 ops/s
# Warmup Iteration   5: 4316518.136 ops/s
Iteration   1: 4311178.994 ops/s
Iteration   2: 4315928.946 ops/s
Iteration   3: 4330921.619 ops/s
Iteration   4: 4331916.972 ops/s
Iteration   5: 4320582.613 ops/s


Result "io.zeebe.journal.file.JournalBenchmark.concurrentReadersWithRWLock":
  4207523.276 ±(99.9%) 106645.131 ops/s [Average]
  (min, avg, max) = (4042818.270, 4207523.276, 4461719.771), stdev = 142368.199
  CI (99.9%): [4100878.145, 4314168.407] (assumes normal distribution)


# Run complete. Total time: 00:08:25

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                                      Mode  Cnt        Score        Error  Units
JournalBenchmark.concurrentReadersWithRWLock  thrpt   25  4207523.276 ± 106645.131  ops/s

```
----------------------------------------------------------------------------------------

```
# JMH version: 1.29
# VM version: JDK 11.0.10, OpenJDK 64-Bit Server VM, 11.0.10+9-Ubuntu-0ubuntu1.20.04
# VM invoker: /usr/lib/jvm/java-11-openjdk-amd64/bin/java
# VM options: -Dvisualvm.id=19206123676576 -javaagent:/home/deepthi/.local/share/JetBrains/Toolbox/apps/IDEA-C/ch-0/211.6693.111/lib/idea_rt.jar=43097:/home/deepthi/.local/share/JetBrains/Toolbox/apps/IDEA-C/ch-0/211.6693.111/bin -Dfile.encoding=UTF-8
# Blackhole mode: full + dont-inline hint
# Warmup: 5 iterations, 10 s each
# Measurement: 5 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 4 threads, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.zeebe.journal.file.JournalBenchmark.concurrentReadersWithStampedLock

# Run progress: 0.00% complete, ETA 00:08:20
# Fork: 1 of 5
# Warmup Iteration   1: 5447395.461 ops/s
# Warmup Iteration   2: 4696760.208 ops/s
# Warmup Iteration   3: 4889011.295 ops/s
# Warmup Iteration   4: 4912055.296 ops/s
# Warmup Iteration   5: 4924511.525 ops/s
Iteration   1: 4851387.498 ops/s
Iteration   2: 4915431.552 ops/s
Iteration   3: 4999947.985 ops/s
Iteration   4: 4868696.729 ops/s
Iteration   5: 4856703.039 ops/s

# Run progress: 20.00% complete, ETA 00:06:45
# Fork: 2 of 5
# Warmup Iteration   1: 4762164.087 ops/s
# Warmup Iteration   2: 4809528.392 ops/s
# Warmup Iteration   3: 4956218.794 ops/s
# Warmup Iteration   4: 4891564.398 ops/s
# Warmup Iteration   5: 4835548.542 ops/s
Iteration   1: 4814378.394 ops/s
Iteration   2: 4797008.719 ops/s
Iteration   3: 4772586.425 ops/s
Iteration   4: 4811166.432 ops/s
Iteration   5: 4788135.287 ops/s

# Run progress: 40.00% complete, ETA 00:05:03
# Fork: 3 of 5
# Warmup Iteration   1: 4671231.985 ops/s
# Warmup Iteration   2: 4645802.879 ops/s
# Warmup Iteration   3: 4636353.557 ops/s
# Warmup Iteration   4: 4629216.485 ops/s
# Warmup Iteration   5: 4585763.726 ops/s
Iteration   1: 4690671.030 ops/s
Iteration   2: 4777969.242 ops/s
Iteration   3: 4619277.721 ops/s
Iteration   4: 4615720.242 ops/s
Iteration   5: 4597631.995 ops/s

# Run progress: 60.00% complete, ETA 00:03:22
# Fork: 4 of 5
# Warmup Iteration   1: 4817619.696 ops/s
# Warmup Iteration   2: 4641540.273 ops/s
# Warmup Iteration   3: 4654258.798 ops/s
# Warmup Iteration   4: 4629749.239 ops/s
# Warmup Iteration   5: 4665599.995 ops/s
Iteration   1: 4594717.052 ops/s
Iteration   2: 4743796.516 ops/s
Iteration   3: 4648604.412 ops/s
Iteration   4: 4659500.757 ops/s
Iteration   5: 4663030.248 ops/s

# Run progress: 80.00% complete, ETA 00:01:41
# Fork: 5 of 5
# Warmup Iteration   1: 4683344.515 ops/s
# Warmup Iteration   2: 4714698.331 ops/s
# Warmup Iteration   3: 4588112.782 ops/s
# Warmup Iteration   4: 4656155.746 ops/s
# Warmup Iteration   5: 4592368.123 ops/s
Iteration   1: 4538534.161 ops/s
Iteration   2: 4616999.821 ops/s
Iteration   3: 4535981.493 ops/s
Iteration   4: 4686109.715 ops/s
Iteration   5: 4636895.038 ops/s


Result "io.zeebe.journal.file.JournalBenchmark.concurrentReadersWithStampedLock":
  4724035.260 ±(99.9%) 91578.724 ops/s [Average]
  (min, avg, max) = (4535981.493, 4724035.260, 4999947.985), stdev = 122254.976
  CI (99.9%): [4632456.537, 4815613.984] (assumes normal distribution)


# Run complete. Total time: 00:08:25

REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                                           Mode  Cnt        Score       Error  Units
JournalBenchmark.concurrentReadersWithStampedLock  thrpt   25  4724035.260 ± 91578.724  ops/s

```

**Comment by npepinpe** [comment]
Keep in mind that `synchronized` is not interruptible, so you might want to try with a simple `ReentrantLock` in this case (which has usually similar performance).

**Comment by deepthidevaki** [comment]
I ran a new benchmark which is more closer to our workload. The benchmark contains 1 writer thread which continuously appends records (which doesn't acquire a lock) and periodically acquire lock to delete segments. Then there are 3 reader threads.

```
Benchmark                                                   Mode  Cnt        Score         Error  Units
JournalBenchmark.concurrentReadersSynchronizedWithWriter   thrpt    5  4343435.193 ±  782597.517  ops/s
JournalBenchmark.concurrentReaderStampedLockWithWriter     thrpt    5  4335469.389 ±  576286.407  ops/s
JournalBenchmark.concurrentReadersReentrantLockWithWriter  thrpt    5  3034155.647 ±  891586.500  ops/s
JournalBenchmark.concurrentReadersRWLockWithWriter         thrpt    5  2296890.969 ± 1124318.610  ops/s
```
From the above results StampedLock seems to be a good choice. The code using StampedLock is also not complex. I would use StampedLock instead of ReentrantReadWriteLock. Note that in our usual cluster benchmark `ReentrantReadWriteLock` did not show any performance impact.
Thanks @npepinpe for pointing this out. It was worth investigating it further.

The following is only for reference and doesn't reflect the real workload.. This benchmark has 4 readers which reads from pre-existing journal. It repeatedly reads the existing records. When the readers reach the end of the journal they restart from the beginning.
```

JournalBenchmark.concurrentReadersSynchronized             thrpt    5  3295240.922 ±  537577.930  ops/s
JournalBenchmark.concurrentReadersStampedLock              thrpt    5  2610964.266 ±  190568.103  ops/s
JournalBenchmark.concurrentReadersRWLock                   thrpt    5  2140452.279 ±   47137.761  ops/s
JournalBenchmark.concurrentReadersReentrantLock            thrpt    5  1741697.811 ±  136062.997  ops/s
```

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java [hunk 1]
```diff
@@ -151,8 +158,14 @@ public void deleteUntil(final long index) {
 
   @Override
   public void reset(final long nextIndex) {
-    journalIndex.clear();
-    writer.reset(nextIndex);
+    final var lock = rwlock.writeLock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java:161]: Same here

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java [hunk 2]
```diff
@@ -45,11 +45,25 @@ public long getNextIndex() {
 
   @Override
   public boolean hasNext() {
-    return hasNextEntry();
+    final var lock = journal.acquireReadlock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java:48]: Same here

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java [hunk 3]
```diff
@@ -45,11 +45,25 @@ public long getNextIndex() {
...
+    try {
+      return unsafeHasNext();
+    } finally {
+      journal.releaseReadlock(lock);
+    }
   }
 
   @Override
   public JournalRecord next() {
+    final var lock = journal.acquireReadlock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java:58]: Same here

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java [hunk 4]
```diff
@@ -65,34 +79,33 @@ public JournalRecord next() {
...
-      seekToFirst();
-    }
-
-    if (index < currentReader.getNextIndex()) {
-      rewind(index);
-    } else if (index > currentReader.getNextIndex()) {
-      forward(index);
-    } else {
-      currentReader.seek(index);
+    final var lock = journal.acquireReadlock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java:82]: Same here

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java [hunk 5]
```diff
@@ -65,34 +79,33 @@ public JournalRecord next() {
...
     }
-
-    return getNextIndex();
   }
 
   @Override
   public long seekToFirst() {
-    replaceCurrentSegment(journal.getFirstSegment());
-    return journal.getFirstIndex();
+    final var lock = journal.acquireReadlock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java:93]: Same here

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java [hunk 6]
```diff
@@ -65,34 +79,33 @@ public JournalRecord next() {
...
+    }
   }
 
   @Override
   public long seekToLast() {
-    replaceCurrentSegment(journal.getLastSegment());
-    seek(journal.getLastIndex());
-
-    return journal.getLastIndex();
+    final var lock = journal.acquireReadlock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java:103]: Same here

File: journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java [hunk 7]
```diff
@@ -102,42 +115,47 @@ public long seekToAsqn(final long asqn) {
...
-    final var index = journalIndex.lookupAsqn(asqn, indexUpperBound);
-
-    // depending on the type of index, it's possible there is no ASQN indexed, in which case start
-    // from the beginning
-    if (index == null) {
-      seekToFirst();
-    } else {
-      seek(index);
-    }
+    final var lock = journal.acquireReadlock();
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/SegmentedJournalReader.java:118]: Same

**Review by miguelpires — CHANGES_REQUESTED** [review]
Looks good. I only have one comment about the stamp naming

**Review by miguelpires — APPROVED** [review]
Thanks! :+1:

**Pull request merged**